R/assocTestSingle.R

```diff
@@ -1,132 +1,151 @@
-          })
-
-
-
-setMethod("assocTestSingle",
-          "GenotypeIterator",
-          function(gdsobj, null.model, test=c("Score", "Wald"), GxE=NULL, male.diploid=TRUE, verbose=TRUE) {
-              test <- match.arg(test)
-
-              # filter samples to match null model
-              sample.index <- .sampleIndexNullModel(gdsobj, null.model)
-              
-              if (!is.null(GxE)) GxE <- .modelMatrixColumns(null.model, GxE)
-              
-              # results
-              res <- list()
-              n.iter <- length(snpFilter(gdsobj))
-              set.messages <- ceiling(n.iter / 100) # max messages = 100
-              i <- 1
-              iterate <- TRUE
-              while (iterate) {
-                  var.info <- variantInfo(gdsobj)
-                  
-                  geno <- getGenotypeSelection(gdsobj, scan=sample.index, order="selection",
-                                               transpose=TRUE, use.names=FALSE, drop=FALSE)
-                  
-                  # take note of number of non-missing samples
-                  n.obs <- colSums(!is.na(geno))
-                  
-                  # allele frequency
-                  freq <- .alleleFreq(gdsobj, geno, sample.index=sample.index,
-                                      male.diploid=male.diploid)
-                  
-                  # filter monomorphic variants
-                  keep <- .filterMonomorphic(geno, count=n.obs, freq=freq$freq)
-                  if (!all(keep)) {
-                      var.info <- var.info[keep,,drop=FALSE]
-                      geno <- geno[,keep,drop=FALSE]
-                      n.obs <- n.obs[keep]
-                      freq <- freq[keep,,drop=FALSE]
-                  }
-
-                  # mean impute missing values
-                  if (any(n.obs < nrow(geno))) {
-                      geno <- .meanImpute(geno, freq$freq)
-                  }
-
-                  # do the test
-                  assoc <- testGenoSingleVar(null.model, G=geno, E=GxE, test=test)
-
-                  res[[i]] <- cbind(var.info, n.obs, freq, assoc)
-                  
-                  if (verbose & n.iter > 1 & i %% set.messages == 0) {
-                      message(paste("Iteration", i , "of", n.iter, "completed"))
-                  }
-                  i <- i + 1
-                  iterate <- GWASTools::iterateFilter(gdsobj)
-              }
-
-              do.call(rbind, res)
-          })
+setGeneric("assocTestSingle", function(gdsobj, ...) standardGeneric("assocTestSingle"))
+
+## do we want the GxE.return.cov option?
+## do we want to make imputing to the mean optional?
+setMethod("assocTestSingle",
+          "SeqVarIterator",
+          function(gdsobj, null.model,  test=c("Score", "Wald",  "SAIGE", "BinomiRare", "CMP"), 
+                   GxE=NULL, sparse=TRUE, imputed=FALSE, male.diploid=TRUE, genome.build=c("hg19", "hg38"), 
+                   AF.max=NULL, calc.score=FALSE, verbose=TRUE) {

```

Comment:
It seems odd to me to have an AF.max argument on single variant tests. What is the motivation behind this?